### ISSUE TITLE
Gateway Timeout

### ISSUE THREAD

**Issue by gusiri** [Consensys/besu-sequencer-plugins#1042]
## Ranges

@gusiri 

```rust
5101295-5101358
5118079-5118105
5118657-5118686
5118971-5119001
5121987-5122023
5122351-5122385
```

@FlorianHuc 

```rust
6871261-6871263
6930360-6930360
7032397-7032402
7032685-7032688
7037237-7037243
7037244-7037244
7037321-7037321
7040245-7040246
```

## Issue

Linea-tracer: v0.5.1-beta
Constraints: v0.5.1-beta (b6bf989)
Corset: latest (fcd3035)
Zkevm.bin: v0.5.1-beta (b6bf989) + HUB, GAS, OOB, and STP disabled

tracer failed but sometimes we get half-generated traces. When we run the checker we get these errors. 

[tracer-failure-gateway-timeout]corset_error_unwrap.tar.gz:
https://drive.google.com/file/d/1BnucvoSprOtXx29iBTWDTvPO55VSecgw/view?usp=drive_link

[tracer-failure-gateway-timeout]corset_error_not_enough_bytes.tar.gz
https://drive.google.com/file/d/1BnucvoSprOtXx29iBTWDTvPO55VSecgw/view?usp=drive_link

----

Note from @FlorianHuc : Regarding the gateway timeout, i would not focus too much on it today. We'll increase the VM size, and we should ask Ameziane to take a look at it.

For `curl`, `--connect-timeout 600` (10 mins) was used for the requests. So it's related to some (possibly, many different) server-side errors or server side timeout limits. 

Some ranges:
```

Gateway Timeout
real    5m0.006s
user    0m0.014s
sys     0m0.001s
[Done]  trace for range 5101295-5101358

Gateway Timeout
real    7m54.318s
user    0m0.022s
sys     0m0.000s
[Done]  trace for range 5118079-5118105

Gateway Timeout
real    5m2.332s
user    0m0.009s
sys     0m0.005s
[Done]  trace for range 5118971-5119001

Gateway Timeout
real    5m2.331s
user    0m0.000s
sys     0m0.014s
[Done]  trace for range 5118657-5118686


Gateway Timeout
real    5m0.552s
user    0m0.015s
sys     0m0.000s
[Done]  trace for range 5121987-5122023

Gateway Timeout
real    5m0.552s
user    0m0.014s
sys     0m0.000s
[Done]  trace for range 5122351-5122385


```

**Comment by FlorianHuc** [comment]
link does not work for me

**Comment by gusiri** [comment]
@FlorianHuc Sorry, just updated the links.

**Comment by FlorianHuc** [comment]
some more ranges that are likely to fail for the same reasons:

"startBlockNumber":6871261,"endBlockNumber":6871263
"startBlockNumber":6930360,"endBlockNumber":6930360
"startBlockNumber":7040245,"endBlockNumber":7040246
"startBlockNumber":7037321,"endBlockNumber":7037321
"startBlockNumber":7037237,"endBlockNumber":7037243
"startBlockNumber":7037244,"endBlockNumber":7037244
"startBlockNumber":7032685,"endBlockNumber":7032688
"startBlockNumber":7032397,"endBlockNumber":7032402

**Comment by gusiri** [comment]
```
Error downloading trace for range 4749485-4749536: Expecting value: line 1 column 1 (char 0)
```


```
Testing range: 4749485 - 4749536
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   245    0    15    0   230      0      0 --:--:--  0:05:00 --:--:--     4
Gateway Timeout

Testing range: 4749485 - 4749536
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   245    0    15    0   230      0      0 --:--:--  0:05:00 --:--:--     4
Gateway Timeout

Testing range: 4749485 - 4749510
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   391    0   161  100   230      1      1  0:03:50  0:02:35  0:01:15    48
{"jsonrpc":"2.0","id":1,"result":{"tracesEngineVersion":"v0.5.3-beta","conflatedTracesFileName":"/data/besu/conflated/4749485-4749510.conflated.v0.5.3-beta.lt"}}
Range 4749485 - 4749510 is successful, skipping further investigation in this range.

Testing range: 4749511 - 4749536
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   391    0   161  100   230      0      1  0:03:50  0:03:04  0:00:46    43
{"jsonrpc":"2.0","id":1,"result":{"tracesEngineVersion":"v0.5.3-beta","conflatedTracesFileName":"/data/besu/conflated/4749511-4749536.conflated.v0.5.3-beta.lt"}}
Range 4749511 - 4749536 is successful, skipping further investigation in this range.
[ec2-user@ip-172-31-6-114 integration_test]$ ./binary_search.sh 4749485 4749536

Testing range: 4749485 - 4749536
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   391    0   161    0   230      0      0 --:--:--  0:04:26 --:--:--    42
{"jsonrpc":"2.0","id":1,"result":{"tracesEngineVersion":"v0.5.3-beta","conflatedTracesFileName":"/data/besu/conflated/4749485-4749536.conflated.v0.5.3-beta.lt"}}
Range 4749485 - 4749536 is successful, skipping further investigation in this range.
```

We get `Gateway Timeout`errors  even when the server is idle, but retrying the same range works. 
It turned out that when you parse `Gateway Timeout` error message using JSON decoder to parse the response, you get `'Expecting value: line 1 column 1 (char 0)'` because the `Gateway Timeout` message is not sent in the response, but directly to your terminal.
So I guess it's the same issue. 
Related issue: Expecting value: line 1 column 1 (char 0)(https://github.com/Consensys/linea-tracer/issues/1069)

**Comment by gusiri** [comment]
I added three retries when downloading the trace, but sometimes three retries fail as well

**Comment by letypequividelespoubelles** [comment]
is it still the case with 0.5.3-beta ?

**Comment by gusiri** [comment]
Yes, I see many `Gateway Timeout`s, but this seems to be related to the CPU/memory resources of the Besu node.

This happens more with ranges from the `mainnet` compared to `sepolia`:
- Size of a .lt file in mainnet: 10+GB
- Size of a .lt file in Sepolia: ~1GB

Also, it occurs more frequently on a Besu node with fewer resources (many Gateway Timeouts on a 32-core, 123GB machine compared to a 64-core, 246GB machine)."

**Comment by gusiri** [comment]
I further investigated and found out what’s going on.

It’s not exactly resource-related. Of course, if we use a machine with more cores and finish the conflation within 5 minutes, we don’t see this error, but that’s not the root cause.

When the size of the trace is too large to be generated within 5 minutes, the Besu RPC drops the request after 5 mins for reasons we don’t yet understand. I used a very long timeout for the HTTP POST request on the client side, so it seems to be a server-side issue.

----

What happens after the Gateway Timeout?

The Besu node keeps generating the trace.
For the range `5068897-5068936.conflated.v0.6.0-rc1.lt`, it took 30 minutes to complete.
When I ran the checker, it passed.

After 1 hour, when I sent another request for the conflation with the same range, it started overwriting the conflation file again, and the client received another `Gateway Timeout` error.
It took another 30 minutes to complete the conflation.
When I ran the checker, it passed.

So, retries don’t help at all, but actually worsen the problem because they just restart the conflation.

----

Temporary solution on the client-side:
I wrote some code to check if the conflation is still ongoing or has been completed by monitoring the remote file.
After the `Gateway Timeout`, it monitors the file and start downloading the file once the conflation is completed. 
Now, I can download the trace regardless of the `Gateway Timeout`.

But we still need to find the root cause. As far as I know, there's no timeout option in the Besu RPC HTTP configuration. It could be somewhere between the plugin and Besu.
